### ISSUE TITLE
rclone serve restic does not close unix sockets cleanly

### ISSUE THREAD

**Issue by ElRico** [rclone/rclone#6648]
#### The associated forum post URL from `https://forum.rclone.org`

https://forum.rclone.org/t/rclone-v1-61-0-release/34914/3?u=el_rico

#### What is the problem you are having with rclone?

`rclone serve restic` does not close unix sockets cleanly, so that subsequent executions of `rclone serve restic` fail, unless the socket is removed from the filesystem beforehand.

#### What is your rclone version (output from `rclone version`)

rclone v1.61.0
- os/version: debian 11.6 (64 bit)
- os/kernel: 5.10.0-20-amd64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.19.4
- go/linking: static
- go/tags: none

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

debian 11.6 (64 bit)

#### Which cloud storage system are you using? (e.g. Google Drive)

Jottacloud

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
```

restic snapshots --verbose -o rclone.args="serve restic --stdio --addr 'unix:///tmp/rclone_test.sock' --b2-hard-delete -vv"
echo "-------"
restic snapshots --verbose -o rclone.args="serve restic --stdio --addr 'unix:///tmp/rclone_test.sock' --b2-hard-delete -vv"
echo "-------"
rm /tmp/rclone_test.sock
restic snapshots --verbose -o rclone.args="serve restic --stdio --addr 'unix:///tmp/rclone_test.sock' --b2-hard-delete -vv
```

#6647 seems to apply here as well, which is why I am using the --stdio option.

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
rclone: 2022/12/21 16:21:07 DEBUG : rclone: Version "v1.61.0" starting with parameters ["rclone" "serve" "restic" "--stdio" "--addr" "unix:///tmp/rclone_test.sock" "--b2-hard-delete" "-vv" "<redacted>"]
rclone: 2022/12/21 16:21:07 DEBUG : Creating backend with remote "<redacted>"
rclone: 2022/12/21 16:21:07 DEBUG : Using config file from "<redacted>"
rclone: 2022/12/21 16:21:07 NOTICE: jottacloud root '<redacted>': Serving restic REST API on []
rclone: 2022/12/21 16:21:08 DEBUG : file-5577006791947779410: GET request error: object not found
rclone: 2022/12/21 16:21:08 DEBUG : keys: list request
repository 46963384 opened (repository version 1) successfully, password is correct
<snip>
-------
rclone: 2022/12/21 16:21:09 DEBUG : rclone: Version "v1.61.0" starting with parameters ["rclone" "serve" "restic" "--stdio" "--addr" "unix:///tmp/rclone_test.sock" "--b2-hard-delete" "-vv" "<redacted>"]
rclone: 2022/12/21 16:21:09 DEBUG : Creating backend with remote "<redacted>"
rclone: 2022/12/21 16:21:09 DEBUG : Using config file from "<redacted>"
rclone: 2022/12/21 16:21:09 INFO  : 
rclone: Transferred:              0 B / 0 B, -, 0 B/s, ETA -
rclone: Errors:                 1 (retrying may help)
rclone: Elapsed time:         0.3s
rclone: 
rclone: 2022/12/21 16:21:09 DEBUG : 5 go routines active
rclone: 2022/12/21 16:21:09 Failed to restic: failed to init server: listen unix /tmp/rclone_test.sock: bind: address already in use
Fatal: unable to open repository at rclone:jottacloud:<redacted>: error talking HTTP to rclone: context canceled
-------
rclone: 2022/12/21 16:21:09 DEBUG : rclone: Version "v1.61.0" starting with parameters ["rclone" "serve" "restic" "--stdio" "--addr" "unix:///tmp/rclone_test.sock" "--b2-hard-delete" "-vv" "<redacted>"]
rclone: 2022/12/21 16:21:09 DEBUG : Creating backend with remote "<redacted>"
rclone: 2022/12/21 16:21:09 DEBUG : Using config file from "<redacted>"
rclone: 2022/12/21 16:21:10 NOTICE: jottacloud root '<redacted>': Serving restic REST API on []
rclone: 2022/12/21 16:21:10 DEBUG : file-5577006791947779410: GET request error: object not found
rclone: 2022/12/21 16:21:10 DEBUG : keys: list request
repository 46963384 opened (repository version 1) successfully, password is correct
<snip>
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by darthShadow** [comment]
I think this may also be due to the duplicate calls to Serve() (https://github.com/rclone/rclone/commit/bfd225166f334b31e22298fca18934a31ff1e27b#commitcomment-93820465).
Will need to wait for ncw to confirm.

**Comment by ncw** [comment]
I've fixed this now. It was caused by missing finalization code in serve restic (and serve webdav) which I've moved into the shared component lib/http.

This is on the same branch as the other restic fixes so will have all the fixes in.

[v1.62.0-beta.6662.1a8e64497.fix-6644-serve-restic](https://beta.rclone.org/branch/fix-6644-serve-restic/v1.62.0-beta.6662.1a8e64497.fix-6644-serve-restic/) on branch [fix-6644-serve-restic](https://github.com/rclone/rclone/tree/fix-6644-serve-restic) (uploaded in 15-30 mins)

@ElRico please test - thanks

**Comment by ncw** [comment]
This fix has been released in v1.61.1 which is available now

**Comment by ElRico** [comment]
This is just to confirm that 1.61.1 does indeed fix this bug on my system. Thank you for the fix @ncw!
